futures-util/src/future/shared.rs

```diff
@@ -156,153 +196,76 @@ impl<F> Future for Shared<F>
                 }
             }
 
-            let _reset = Reset(&self.inner.notifier.state);
+            let _reset = Reset(&this.inner.notifier.state);
 
             // Poll the future
             let res = unsafe {
-                let (ref mut f, ref mut data) = *(*self.inner.future.get()).as_mut().unwrap();
-                let waker = Waker::from(self.inner.notifier.clone());
-                let mut cx = task::Context::new(data, &waker, cx.executor());
-                f.poll(&mut cx)
+                if let FutureOrOutput::Future(future) = &mut *this.inner.future_or_output.get() {
+                    PinMut::new_unchecked(future).poll(&mut cx)
+                } else {
+                    unreachable!()
+                }
             };
             match res {
-                Ok(Async::Pending) => {
+                Poll::Pending => {
                     // Not ready, try to release the handle
-                    match self.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
+                    match this.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
                         POLLING => {
                             // Success
-                            return Ok(Async::Pending);
+                            return Poll::Pending;
                         }
                         REPOLL => {
                             // Gotta poll again!
-                            let prev = self.inner.notifier.state.swap(POLLING, SeqCst);
+                            let prev = this.inner.notifier.state.swap(POLLING, SeqCst);
                             assert_eq!(prev, REPOLL);
                         }
                         _ => unreachable!(),
                     }
-
-                }
-                Ok(Async::Ready(i)) => {
-                    unsafe {
-                        (*self.inner.result.get()) = Some(Ok(SharedItem { item: Arc::new(i) }));
-                    }
-
-                    break;
                 }
-                Err(e) => {
+                Poll::Ready(output) => {
+                    let output = Arc::new(output);
                     unsafe {
-                        (*self.inner.result.get()) = Some(Err(SharedError { error: Arc::new(e) }));
+                        *this.inner.future_or_output.get() =
+                            FutureOrOutput::Output(output.clone());
                     }
-
-                    break;
+                    // Complete the future
+                    this.inner.notifier.state.store(COMPLETE, SeqCst);
+                    Wake::wake(&this.inner.notifier);
+                    return Poll::Ready(output);
```

Comment:
Here is where I suggest to destroy the slab

Reply:
It will get immediately reset if done there because `Shared::poll` sets the `waker` before doing anything else (like checking if the value is complete). We could set COMPLETE while holding onto the `notifiers` lock, and check it before amending after taking out the lock in `Shared::poll`-- I'll give that a shot.

Reply:
set_waker could be a nop if *the option* is `None` :)

Reply:
Done.

Reply:
Bene, bene 👌